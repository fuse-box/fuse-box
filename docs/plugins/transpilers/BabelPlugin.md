# BabelPlugin

## Description
The babel plugin is used to transpile code to different dialects of javascript.

## Install
This package depends on the `babel-core` module.

```bash
yarn add babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --dev
npm install babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --save-dev
```

##### Parameters

| Name | Type | Description | ?Default |
| ---- | ---- | ----------- | -------- |
| config | `Object`  | when using other fuse-box only properties, babel config is passed in as .config |  |
| limit2project | `boolean`  | to use this plugin across an entire project (including other modules like npm) | `true` |
| extensions | `Array<string>`  | file extensions to allow with fuse-box | `[".jsx"]`
| test | `Regex`  | files to match | <code>/\\.(j&#124;t)s(x)?$/</code> |




## Examples

### JSX

```js
const {BabelPlugin} = require("fuse-box");

plugins: [
    BabelPlugin({
      config: {
        sourceMaps: true,
        presets: ["es2015"],
        plugins: [
            ["transform-react-jsx"],
        ],
      },
    })
]
```

### Shorthand

when only passing in a babel config, there is no need for .config

```js
plugins: [
  BabelPlugin({
    presets: ["es2015"]
  })
]
```
