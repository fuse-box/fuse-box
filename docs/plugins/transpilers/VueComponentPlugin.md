# VueComponentPlugin

## Description
The VueComponentPlugin is used to transpile `.vue` files into Javascript.

## Install
This plugin depends on the following node modules:
- `vue-template-compiler`
- `vue-template-es2015-compiler`
- `vue-hot-reload-api`
- `postcss-selector-parser`
- `vue`

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --save-dev
```

For more information about the `*.vue` file specifications please see the official [documentation](https://vue-loader.vuejs.org/en/start/spec.html)

## Usage
Just import from FuseBox:
```js
const { VueComponentPlugin } = require('fuse-box')
```

Inject into a chain:
```js
fuse.plugin(
  VueComponentPlugin()
)
```

Or add it to the main config to make it available across bundles:
```js
FuseBox.init({
  plugins: [
    VueComponentPlugin()
  ]
});
```

## Configuration
### Defaults
If no `lang` attributes are specified (see below) the `VueComponentPlugin` will use the following configuration out of the box:

- `<template>` Defaults to `html` and uses [HTMLPlugin](/plugins/html-plugin)
- `<script>` Defaults to `ts` and uses FuseBox's integrated Typescript compiler
- `<style>` Defaults to `css` and uses [CSSPlugin](/plugins/css-plugin)

### Language Attributes
When `VueComponentPlugin` detects a `lang` attribute on a block it will attempt to match with the corresponding FuseBox plugin. For example:

- `<script lang="coffee">` - [CoffeeScriptPlugin](/plugins/coffeescript-plugin)
- `<script lang="js">` - [BabelPlugin](/plugins/babelplugin) (will read configuration from `.babelrc` file)
- `<style lang="less">` - [LESSPlugin](/plugins/less-plugin)
- `<style lang="scss">` - [SASSPlugin](/plugins/sass-plugin)

note: If using lang="js" and configuration item useTypescriptCompiler: true then FuseBox will use the internal Typescript compiler and NOT BabelPlugin


### Using Custom Plugin Chains
If the above functionality doesn't fit your needs, you can override the pre-processing by optionally setting the `script`, `style` or `template` options. This follows the standard FuseBox way of defining plugin chains:

```js
const { FuseBox, VueComponentPlugin, BabelPlugin, SassPlugin, CSSResourcePlugin, PostCSSPlugin, CSSPlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
      VueComponentPlugin({
          script: BabelPlugin({ ... }),
          template: HTMLPlugin({ ... })
          style: [
              SassPlugin({ ... }),
              CSSResourcePlugin({ ... }),
              PostCSSPlugin( ... ),
              CSSPlugin({ ... })
          ]
      })
    ]
})
```

note: Overriding a plugin chain for a .vue block will make the VueComponentPlugin ignore any lang attributes.

### External Files
The `VueComponentPlugin` fully understands the `src` attribute and will handle external files just the same as inline content.

### Scoped Styling
Scoped styling is fully supported by using the `scoped` attribute. Support for the `module` attribute will be coming soon.

### HMR
Hot Module Reloading is also fully supported, just enable it on your bundles by calling `.hmr()`. See [here](/page/development#hot-module-reload) for more info.
