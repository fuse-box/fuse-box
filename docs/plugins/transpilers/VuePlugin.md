# VuePlugin

## Description
The Vue plugin is used to transpile .vue files into Javascript.

## Install
This package depends on the `vue-template-compiler`, `vue-template-es2015-compiler` and one of `typescript` and `babel-core` modules.

Also, if you want to include Vue library into the bundle, remember to install `vue` module too.

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --save-dev
```

## Example

```js
// fuse.js

const { FuseBox, VuePlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
        VuePlugin()
    ]
})

fsbx.dev()

fsbx.bundle('app.js')
    .instructions('>index.ts')
    .watch()

fsbx.run()

```

```typescript
// src/index.ts

import * as Vue from 'vue/dist/vue.js'
import * as App from './components/App.vue'

new Vue({
    el: '#app',
    render: h => h(App)
})

```

```html
<!-- src/components/App.vue -->

<template>
    <div>
        <p>{{ msg }}</p>
        <input type="text" v-model="msg" />
    </div>
</template>

<script>
    export default {
        name: 'app',
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```

## Script transpiler

By default `VuePlugin` uses `typescript` to compile contents of `<script></script>`. By defining the `lang` attribute of the `<script>` tag, content can be compiled using desired transpiler. Currently only `ts` and `babel` are supported.

### Babel

If `babel` attribute is given, VuePlugin will load babel config from there, otherwise VuePlugin will load `.babelrc` from the root of the project. If no config given and no rc found, it uses default config. Packages `babel-core` and `babel-plugin-transform-es2015-modules-commonjs` must be installed to use babel.

```js
// Configuring babel in fuse.js (Default)
VuePlugin({
    babel: {
        config: {
            "plugins": ["transform-es2015-modules-commonjs"]
        }
    }
})

// .babelrc
{
 "plugins": ["transform-es2015-modules-commonjs"]
}
```

## Template engine preprocessor

By defining the `lang` attribute of the `<template>` tag, content can be compiled to HTML using your favorite template engine.

The `consolidate` module must be installed in addition to the template engine module you want to use.

View a list of all [supported template engines](https://github.com/tj/consolidate.js#supported-template-engines).

### Example using Pug:

```bash
# Using yarn:
yarn add consolidate pug --dev

# Using npm:
npm install consolidate pug --save-dev
```

```html
<!-- src/components/App.vue -->

<template lang="pug">
    .content
        p {{ msg }}
        input(type="text", v-model="msg")
</template>

<script>
    export default {
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```
