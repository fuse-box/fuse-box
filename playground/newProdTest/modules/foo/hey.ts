export function Hey() {
  return 'Hey';
}
