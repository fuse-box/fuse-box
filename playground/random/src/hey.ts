export interface BInterface {}
export interface AInterface {
  b?: BInterface;
}

const AObject: AInterface = {
  b: {},
};
export { AObject };
