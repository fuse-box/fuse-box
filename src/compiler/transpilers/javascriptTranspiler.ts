import { ASTNode } from '../interfaces/AST';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { ITranspiler } from '../interfaces/ITranspiler';
import { createGlobalContext } from '../program/GlobalContext';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BundlePolyfillTransformer } from '../transformers/bundle/BundlePolyfillTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ITransformerResult } from '../interfaces/ITranformerResult';

export function javascriptTranspiler(props: ITranspiler): ITransformerResult {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: props.env,
      isBrowser: props.isBrowser,
      isServer: props.isServer,
    }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    props.target === 'browser' && BrowserProcessTransformer({ env: props.env, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...props, onRequireCallExpression }),

    DynamicImportTransformer({ onRequireCallExpression }),

    JSXTransformer(),

    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  if (props.transformers && props.transformers.length) {
    const opts = { module: props.module, onRequireCallExpression };
    for (const transformer of props.transformers) {
      defaultTransformers.unshift(transformer(opts));
    }
  }

  transpileModule({
    ast: props.ast as ASTNode,
    globalContext: createGlobalContext({}),
    transformers: defaultTransformers,
  });

  return { ast: props.ast, requireStatementCollection };
}
