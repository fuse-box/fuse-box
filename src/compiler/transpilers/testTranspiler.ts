import * as fs from 'fs';
import * as path from 'path';
import { ITarget } from '../../config/PrivateConfig';
import { createContext } from '../../core/Context';
import { createModule } from '../../core/Module';
import { createPackage } from '../../core/Package';
import { makeFuseBoxPath } from '../../utils/utils';
import { generate } from '../generator/generator';
import { ASTNode } from '../interfaces/AST';
import { parseTypeScript, parseJavascript } from '../parser';
import { transformCommonVisitors } from '../transformer';

export interface ICompileModuleProps {
  target?: ITarget;
  env?: { [key: string]: string };
  fileName?: string;
  code?: string;
  emitDecoratorMetadata?: boolean;
  useMeriyah?: boolean;
}

export function testTranspile(props: ICompileModuleProps) {
  let contents = props.code;

  if (!contents) {
    contents = fs.readFileSync(props.fileName).toString();
  }

  const ext = props.fileName ? path.extname(props.fileName) : '.tsx';

  let ast;
  if (props.useMeriyah) {
    ast = parseJavascript(contents, { jsx: true });
  } else {
    ast = parseTypeScript(contents, { jsx: true });
  }

  const ctx = createContext({
    cache: false,
    devServer: false,
    target: props.target || 'browser',
    env: props.env,
  });

  if (props.emitDecoratorMetadata) {
    ctx.tsConfig.compilerOptions.emitDecoratorMetadata = props.emitDecoratorMetadata;
  }

  const pkg = createPackage({ ctx: ctx, meta: {} as any });
  const module = createModule(
    {
      absPath: __filename,
      ctx: ctx,
      extension: ext,
      fuseBoxPath: makeFuseBoxPath(path.dirname(props.fileName || __filename), props.fileName || __filename),
    },
    pkg,
  );
  module.ast = ast as ASTNode;

  transformCommonVisitors(module);

  const res = generate(ast, {});

  return res;
}
