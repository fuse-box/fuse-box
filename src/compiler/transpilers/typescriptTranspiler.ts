import { ASTNode } from '../interfaces/AST';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { ITranspiler } from '../interfaces/ITranspiler';
import { createGlobalContext } from '../program/GlobalContext';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BundlePolyfillTransformer } from '../transformers/bundle/BundlePolyfillTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ClassConstructorPropertyTransformer } from '../transformers/ts/ClassConstructorPropertyTransformer';
import { CommonTSfeaturesTransformer } from '../transformers/ts/CommonTSfeaturesTransformer';
import { EnumTransformer } from '../transformers/ts/EnumTransformer';
import { NamespaceTransformer } from '../transformers/ts/NameSpaceTransformer';
import { ITransformerResult } from '../interfaces/ITranformerResult';
import { DecoratorTransformer } from '../transformers/ts/decorators/DecoratorTransformer';

export function typescriptTranspiler(props: ITranspiler): ITransformerResult {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: props.env,
      isBrowser: props.isBrowser,
      isServer: props.isServer,
    }),
    DecoratorTransformer({ onRequireCallExpression, emitDecoratorMetadata: props.emitDecoratorMetadata }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    props.target === 'browser' && BrowserProcessTransformer({ env: props.env, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...props, onRequireCallExpression }),

    DynamicImportTransformer({ onRequireCallExpression }),
    EnumTransformer(),
    ClassConstructorPropertyTransformer(),
    JSXTransformer(),
    NamespaceTransformer(),

    CommonTSfeaturesTransformer(),
    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  if (props.transformers && props.transformers.length) {
    const opts = { module: props.module, onRequireCallExpression };
    for (const transformer of props.transformers) {
      defaultTransformers.unshift(transformer(opts));
    }
  }

  transpileModule({
    ast: props.ast as ASTNode,
    globalContext: createGlobalContext({}),
    transformers: defaultTransformers,
  });

  return { ast: props.ast, requireStatementCollection };
}
