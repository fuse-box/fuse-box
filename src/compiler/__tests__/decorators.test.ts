import { testTranspile } from '../transpilers/testTranspiler';

describe('Decorators test', () => {
  describe('Class decorators', () => {
    it('Should inject 1 decorator ', () => {
      const res = testTranspile({
        code: `
        @hey
        class Foo {}
    `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('Should inject 2 decorators ', () => {
      const res = testTranspile({
        code: `
        @foo
        @bar()
        class Foo {}
    `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('Should inject a decorator on export class', () => {
      const res = testTranspile({
        code: `
        @bar()
        export class Foo {}
    `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('Should inject a decorator on export default class with a name', () => {
      const res = testTranspile({
        code: `
        @bar()
        export default class Foo {}
    `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('Should inject a decorator on export default class without a name', () => {
      const res = testTranspile({
        code: `
        @bar()
        export default class {}
    `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('Should inject inside a default function', () => {
      const res = testTranspile({
        code: `
       export default function(){
        @bar()
        class SomeClass {}
       }
    `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('Should consider import variables too', () => {
      const res = testTranspile({
        code: `
       import hey from "./mod";
       export default function(){
        @bar({ hey })
        class SomeClass {}
       }
    `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('Should have no problems transforming further', () => {
      const res = testTranspile({
        code: `
        @bar()
        export default class {
          public name : string = "Bob"
        }
    `,
      });

      expect(res.code).toMatchSnapshot();
    });
  });

  describe('Class Properties', () => {
    it('should decorate name 1', () => {
      const res = testTranspile({
        code: `
        class A {
          @foo
          public name : string;
        }
    `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('should decorate name 2', () => {
      const res = testTranspile({
        code: `
        class A {
          @foo
          public name : string = "hey";
        }
    `,
      });
      expect(res.code).toMatchSnapshot();
    });
  });

  describe('Method decorators', () => {
    it('should decorate class method', () => {
      const res = testTranspile({
        code: `
        class A {
          @foo
          public hey(){}
        }
    `,
      });
      expect(res).toMatchSnapshot();
    });

    it('should decorate class static method', () => {
      const res = testTranspile({
        code: `
        class A {
          @foo
          public static hey(){}
        }
    `,
      });

      expect(res).toMatchSnapshot();
    });
  });

  describe('Method param decorators', () => {
    it('should contain 1 decorator ', () => {
      const res = testTranspile({
        code: `
        class A {

          hey(@foo a : string){}
        }
    `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('should contain 2 decorators ', () => {
      const res = testTranspile({
        code: `
      class A {

        hey(@foo @bar() a : string){}
      }
  `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('should contain 3 decorators ', () => {
      const res = testTranspile({
        code: `
      class A {

        hey(@foo @bar() a : string, @oi b : string){}
      }
  `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('should contain 1 decorator with a static method ', () => {
      const res = testTranspile({
        code: `
      class A {

        static hey(@foo a : string){}
      }
  `,
      });

      expect(res.code).toMatchSnapshot();
    });
  });

  describe('Metadata', () => {
    describe('Constructor metadata', () => {
      it('should handle normal properties', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          @a
          @b()
          export class HomeComponent {
            constructor(env: Env, bar: Runkari) {}
          }
    `,
        });
        expect(res.code).toMatchSnapshot();
      });

      it('should handle properties with visibility', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          @a
          @b()
          export class HomeComponent {
            constructor(public env: Env, bar) {}
          }
    `,
        });

        expect(res.code).toMatchSnapshot();
      });

      it('should respect imports', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          import Env from "./some-data";
          @a
          @b()
          export class HomeComponent {
            constructor(public env: Env, bar) {}
          }
    `,
        });

        expect(res.code).toMatchSnapshot();
      });

      it('should work ok without any props', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          @a
          @b()
          export class HomeComponent {
            constructor() {}
          }
    `,
        });
        expect(res.code).toMatchSnapshot();
      });

      it('should work ok without constructor', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          @a
          @b()
          export class HomeComponent {

          }
    `,
        });

        expect(res.code).toMatchSnapshot();
      });
    });

    describe('Paramater decorators', () => {
      it('should add 1', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          export class HomeComponent {
            @hey
            public name: string;
          }

    `,
        });
        expect(res.code).toMatchSnapshot();
      });

      it('should add 2 decorators', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          export class HomeComponent {
            @hey
            @oi({})
            public name: string;
          }

    `,
        });
        expect(res.code).toMatchSnapshot();
      });
    });

    describe('method params meta', () => {
      it('should handle 1', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          class Application {
            @hey
            @gay
            name(@kukka @sukka kakka: string) {}
          }
    `,
        });
        expect(res.code).toMatchSnapshot();
      });

      it('should handle with return type', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          class Application {
            @hey
            @gay
            name(@kukka @sukka kakka: string) : HelloSomeObject {}
          }
    `,
        });

        expect(res.code).toMatchSnapshot();
      });

      it('should handle 2', () => {
        const res = testTranspile({
          emitDecoratorMetadata: true,
          code: `
          class Application {
            @oi
            name(some: string): HelloSomeObject {}
          }
    `,
        });

        expect(res.code).toMatchSnapshot();
      });
    });
  });
});
