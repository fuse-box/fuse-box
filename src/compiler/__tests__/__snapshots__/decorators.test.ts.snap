// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decorators test Class Properties should decorate name 1 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {}
__fuse_decorate.d([foo, __fuse_decorate.m(\\"design:type\\", String)], A.prototype, \\"name\\", void 0);
"
`;

exports[`Decorators test Class Properties should decorate name 2 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {
  constructor() {
    this.name = \\"hey\\";
  }
}
__fuse_decorate.d([foo, __fuse_decorate.m(\\"design:type\\", String)], A.prototype, \\"name\\", void 0);
"
`;

exports[`Decorators test Class decorators Should consider import variables too 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
var mod_1 = require(\\"./mod\\");
function __DefaultExport__() {
  class SomeClass {}
  SomeClass = __fuse_decorate.d([bar({
    hey: mod_1.default
  })], SomeClass);
}
exports.default = __DefaultExport__;
"
`;

exports[`Decorators test Class decorators Should have no problems transforming further 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let __DefaultExport__ = class __DefaultExport__ {
  constructor() {
    this.name = \\"Bob\\";
  }
};
__DefaultExport__ = __fuse_decorate.d([bar()], __DefaultExport__);
exports.default = __DefaultExport__;
"
`;

exports[`Decorators test Class decorators Should inject 1 decorator  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class Foo {}
Foo = __fuse_decorate.d([hey], Foo);
"
`;

exports[`Decorators test Class decorators Should inject 2 decorators  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class Foo {}
Foo = __fuse_decorate.d([foo, bar()], Foo);
"
`;

exports[`Decorators test Class decorators Should inject a decorator on export class 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let Foo = class Foo {};
Foo = __fuse_decorate.d([bar()], Foo);
exports.Foo = Foo;
"
`;

exports[`Decorators test Class decorators Should inject a decorator on export default class with a name 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let Foo = class Foo {};
Foo = __fuse_decorate.d([bar()], Foo);
exports.default = Foo;
"
`;

exports[`Decorators test Class decorators Should inject a decorator on export default class without a name 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let __DefaultExport__ = class __DefaultExport__ {};
__DefaultExport__ = __fuse_decorate.d([bar()], __DefaultExport__);
exports.default = __DefaultExport__;
"
`;

exports[`Decorators test Class decorators Should inject inside a default function 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
function __DefaultExport__() {
  class SomeClass {}
  SomeClass = __fuse_decorate.d([bar()], SomeClass);
}
exports.default = __DefaultExport__;
"
`;

exports[`Decorators test Metadata Constructor metadata should handle normal properties 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let HomeComponent = class HomeComponent {
  constructor(env, bar) {}
};
HomeComponent = __fuse_decorate.d([a, b(), __fuse_decorate.m(\\"design:paramtypes\\", [typeof Env === \\"function\\" ? Env : Object, typeof Runkari === \\"function\\" ? Runkari : Object])], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Constructor metadata should handle overload gracefully 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
var some_data_1 = require(\\"./some-data\\");
let HomeComponent = class HomeComponent {
  constructor(env, bar) {
    this.env = env;
  }
};
HomeComponent = __fuse_decorate.d([a, b(), __fuse_decorate.m(\\"design:paramtypes\\", [typeof some_data_1.default === \\"function\\" ? some_data_1.default : Object, Object])], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Constructor metadata should handle properties with visibility 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let HomeComponent = class HomeComponent {
  constructor(env, bar) {
    this.env = env;
  }
};
HomeComponent = __fuse_decorate.d([a, b(), __fuse_decorate.m(\\"design:paramtypes\\", [typeof Env === \\"function\\" ? Env : Object, Object])], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Constructor metadata should respect imports 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
var some_data_1 = require(\\"./some-data\\");
let HomeComponent = class HomeComponent {
  constructor(env, bar) {
    this.env = env;
  }
};
HomeComponent = __fuse_decorate.d([a, b(), __fuse_decorate.m(\\"design:paramtypes\\", [typeof some_data_1.default === \\"function\\" ? some_data_1.default : Object, Object])], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Constructor metadata should work ok without any props 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let HomeComponent = class HomeComponent {
  constructor() {}
};
HomeComponent = __fuse_decorate.d([a, b(), __fuse_decorate.m(\\"design:paramtypes\\", [])], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Constructor metadata should work ok without constructor 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
let HomeComponent = class HomeComponent {};
HomeComponent = __fuse_decorate.d([a, b()], HomeComponent);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Paramater decorators should add 1 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class HomeComponent {}
__fuse_decorate.d([hey, __fuse_decorate.m(\\"design:type\\", String)], HomeComponent.prototype, \\"name\\", void 0);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata Paramater decorators should add 2 decorators 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class HomeComponent {}
__fuse_decorate.d([hey, oi({}), __fuse_decorate.m(\\"design:type\\", String)], HomeComponent.prototype, \\"name\\", void 0);
exports.HomeComponent = HomeComponent;
"
`;

exports[`Decorators test Metadata method params meta should handle 1 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class Application {
  name(kakka) {}
}
__fuse_decorate.d([hey, gay, __fuse_decorate.p(0, kukka), __fuse_decorate.p(0, sukka), __fuse_decorate.m(\\"design:type\\", Function), __fuse_decorate.m(\\"design:paramtypes\\", [String]), __fuse_decorate.m(\\"design:returntype\\", void 0)], Application.prototype, \\"name\\", null);
"
`;

exports[`Decorators test Metadata method params meta should handle 2 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class Application {
  name(some) {}
}
__fuse_decorate.d([oi, __fuse_decorate.m(\\"design:type\\", Function), __fuse_decorate.m(\\"design:paramtypes\\", [String]), __fuse_decorate.m(\\"design:returntype\\", typeof HelloSomeObject === \\"function\\" ? HelloSomeObject : Object)], Application.prototype, \\"name\\", null);
"
`;

exports[`Decorators test Metadata method params meta should handle with return type 1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class Application {
  name(kakka) {}
}
__fuse_decorate.d([hey, gay, __fuse_decorate.p(0, kukka), __fuse_decorate.p(0, sukka), __fuse_decorate.m(\\"design:type\\", Function), __fuse_decorate.m(\\"design:paramtypes\\", [String]), __fuse_decorate.m(\\"design:returntype\\", typeof HelloSomeObject === \\"function\\" ? HelloSomeObject : Object)], Application.prototype, \\"name\\", null);
"
`;

exports[`Decorators test Method decorators should decorate class method 1`] = `
Object {
  "code": "class A {
  hey() {}
}
",
  "requireStatementCollection": Array [],
}
`;

exports[`Decorators test Method decorators should decorate class static method 1`] = `
Object {
  "code": "class A {
  static hey() {}
}
",
  "requireStatementCollection": Array [],
}
`;

exports[`Decorators test Method param decorators should contain 1 decorator  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {
  hey(a) {}
}
__fuse_decorate.d([__fuse_decorate.p(0, foo)], A.prototype, \\"hey\\", null);
"
`;

exports[`Decorators test Method param decorators should contain 1 decorator with a static method  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {
  static hey(a) {}
}
__fuse_decorate.d([__fuse_decorate.p(0, foo)], A, \\"hey\\", null);
"
`;

exports[`Decorators test Method param decorators should contain 2 decorators  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {
  hey(a) {}
}
__fuse_decorate.d([__fuse_decorate.p(0, foo), __fuse_decorate.p(0, bar())], A.prototype, \\"hey\\", null);
"
`;

exports[`Decorators test Method param decorators should contain 3 decorators  1`] = `
"var __fuse_decorate = require(\\"fuse_helpers_decorate\\");
class A {
  hey(a, b) {}
}
__fuse_decorate.d([__fuse_decorate.p(0, foo), __fuse_decorate.p(0, bar()), __fuse_decorate.p(1, oi)], A.prototype, \\"hey\\", null);
"
`;
